### ISSUE TITLE
add ffmpeg downloading for setup

### ISSUE THREAD

**Issue by gedw99** [benpate/mediaserver#48]
Will make it easier for devs and users.

https://ffmpeg.org/download.html

Basically it checks for existence of the right OS / ARCH / VERSION, and if not there will download it.

so it can be used at compile or runtime.

It will help make https://github.com/EmissarySocial/emissary setup easier too.

**Comment by gedw99** [comment]
https://github.com/stashapp/stash/blob/develop/pkg/ffmpeg/downloader.go is a decent example I just found...

**Comment by benpate** [comment]
Howdy @gedw99 - thanks for the suggestion.  FFmpeg is pretty new in this stack -- I've only recently gotten it integrated into Emissary, and am actually just posting it to my own production site now. So yes, the documentation around it could be *much* better.

One thing I *have* done is to include FFmpeg in a Docker configuration.  I'm getting an example together for this pretty soon, and it should help those who just want to "kick the tires" a bit.

In addition, I've updated the [ installation page](https://emissary.dev/installation) to point directly to the download URL you gave.  Thank you. That should be easier.

I'm not sure that we should bake download locations directly into the software. Especially for production environments, this feels like a configuration setting that admins should control, and that the app itself should not try to do automatically.

Does this all make sense? Does it work for you?

**Comment by gedw99** [comment]
How about then we embed the ffmpeg at compile time and spit it to disk at startup time. Its simple and clean and no security risks.

```go

package main

import (
    "fmt"
    "io/fs"
    "embed"
)

//go:embed ffmpeg.exe  ffprobe.exe
var x embed.FS

func main() {

    // Walk all files in the embedded file system.
    fs.WalkDir(x, ".", func(path string, d fs.DirEntry, err error) error {
        // Ignore error.
        if err != nil {
            return nil
        }
        // Skip period path.
        if path == "." {
            return nil
        }
        fmt.Println("PATH:", path)

        // Read file.
        content, _ := fs.ReadFile(x, path)

        // Print file length and file if it is short.
        fmt.Println("FILE LENGTH:", len(content))
        if len(content) <= 50 {
            fmt.Println("FILE CONTENTS:", string(content))
        }
        return nil
    })
}

```

**Comment by benpate** [comment]
go-ffstatic looks like an interesting project, but I'm concerned about embedding a whole binary (or several, it seems) into the Emissary download.  It seems like it could be brittle. More, this library hasn't been updated in a couple of years. Does this mean that it's embedding two-year-old FFmpeg?

I've just switched from an embedded CGo library over to FFmpeg, and I'm enjoying the much simpler and quicker compiles.  So, I'd like to let this use of FFmpeg bake a bit before I make any more big changes to how it's used.  I'm expecting it will take on larger responsibilities once I get into HLS video streaming. 

By then, we should have more data about how elegant or cumbersome the current integration is.  I'm hoping that the Docker configuration (or possibly an `apt-get` command directly in the app initialization) will be enough for most environments so that we don't have to add several more binaries into the download package.

**Comment by gedw99** [comment]
I agree with your points . 

half way house then could be:

The external binaries can be just pulled at dev time using a golang or makefile . It at least gives a form of version control.

At deploy time, just put that ffmpeg binary inside the docker. Then you have immutable infrastructure at least.

**Comment by benpate** [comment]
Yeah, I see the benefit of "immutable infrastructure."  Right now, the Docker config just uses `apt-get` to download "the most recent version" of FFmpeg.  This has some advantages but could also be a problem down the road.

I'm only getting started with Docker, so I'm not sure how to "put that ffmpeg binary inside the docker" at deploy time.  I'm imagining it as downloading a known version from a trusted source, depending on OS and CPU architecture.  For instance, it seems reasonable to use `golang:1.22.3-bookworm` as the pre-built Docker image for the base operating system.  Might there be similar, pre-built FFmpeg images that we could use from Docker Hub?

BTW.. Thanks for digging in to this with me.  I very much appreciate your interest in making Emissary work better.  So, thank you!

**Comment by gedw99** [comment]
Its just the COPY command in docker file.

Hence why if the ffmpeg is already local then your just then doing a COPY in docker and its job done.

Thats why I suggested a local dev tool in the first issue item that already does that part for ypu.

we are back to square one :)

feel free to ask any more questions. I know its hard to see collaborate around architecture decisions sometimes.

**Comment by gedw99** [comment]
Once this is done it will be easy to make a fly.yaml file so anyone can deploy to fly.io for server less

**Comment by benpate** [comment]
Wow, this looks pretty useful for managing FFmpeg!  There should be some good way to use this for the installation.

"You will likely hate it..." 😀 No, I don't have it, and it could be a cool part of the setup process.  If FFmpeg doesn't already exist, then try to install it.  I'm working on a Mac (plus Docker) and so I hesitate to take responsibility for all the platforms and FFmpeg versions that I can't directly test.

Could we link to this script somehow in the setup instructions? There are so many different server environments, so maybe we add a flag to the Emissary arguments so that server owners can choose if the setup installs FFmpeg or not?

Also, I am probably putting together an FFmpeg library soon... moving the existing command line wrappers into a set of functions that are a more specific.  Perhaps we make this a part of that?
